### DISCUSSION TITLE
Indent while pasting

### DISCUSSION THREAD

**Discussion by Set27** [vscode-neovim/vscode-neovim#1777] [Q&A]
I don't get how indent while pasting work (or how I can configure it)
For example, O (or o) works perfect, but pasting always past in beginning of string, and I have to tab it (or use formatter each time).
I tried to configure it in `init.vim` and in vscode setting, but nothing help. Furthermore, I also record video with empty config.

https://github.com/vscode-neovim/vscode-neovim/assets/49077110/05a0c931-58b2-4225-a4d1-7ca50d54afe6

**Comment by gaoqiangks** [comment]
I got a solution. 

On neovim side, add the following code into your `init.lua`

```

function yank_paste_select(arg) 
    if arg == "above" then
        vim.api.nvim_exec2(vim.api.nvim_replace_termcodes("normal O<Esc>p", true, false, true), {output=false})
    else
        vim.api.nvim_exec2(vim.api.nvim_replace_termcodes("normal o<Esc>p", true, false, true), {output=false})
    end
    --选中刚刚粘贴的内容
    vim.cmd.normal("`[v`]")
end

keyset("n", "<A-p>", "<cmd>lua yank_paste_select('below')<CR>", { silent = true, remap=false})
keyset("n", "<A-P>", "<cmd>lua yank_paste_select('above')<CR>", { silent = true, remap=false})
```

The above code maps `<alt-p>` to paste yanked text below the current line, and maps `<alt-p>` to paste yanked text above the current line, and then select pasted line.


Install the plugin `multi-commands` on your vscode, and add the following into your `settings.json`.

```
    "multiCommand.commands": [
        {
            "command": "multiCommand.PasteBelowAndIndent",
            "interval": 300,
            "sequence": [
                {
                    "command":"vscode-neovim.send",
                    "args": "<a-p>", 
                },
                "editor.action.reindentselectedlines",
            ]
        },
        {
            "command": "multiCommand.PasteAboveAndIndent",
            "interval": 300,
            "sequence": [
                {
                    "command":"vscode-neovim.send",
                    "args": "<a-P>", 
                },
                "editor.action.reindentselectedlines",
            ]
        },
        {
            "command": "multiCommand.swapChar",
            "interval": 30,
            "sequence": [
                "cursorLeftSelect",
                "editor.action.clipboardCutAction",
                "cursorRight",
                "editor.action.clipboardPasteAction"
            ]
        }
    ]


```

Add the following code into your `keyboard.json`

```
    {
        "key":"alt+p",
        "command": "extension.multiCommand.execute",
        "args": { 
            "sequence":[
                "multiCommand.PasteBelowAndIndent" ,
                {
                    "command":"vscode-neovim.send",
                    "args": "<Esc>",
                }
            ]
        },
        "when": "editorTextFocus && neovim.init" 
    },
    {
        "key":"alt+shift+p",//如果映射的是alt-P的话, alt shift p不起作用, 不知道为什么
        "command": "extension.multiCommand.execute",
        "args": { 
            "sequence":[
                "multiCommand.PasteAboveAndIndent" ,
                {
                    "command":"vscode-neovim.send",
                    "args": "<Esc>",
                }
            ]
        },
        "when": "editorTextFocus && neovim.init" 
    },
```

Now you can use `<alt-p>` to paste yanked lines just below the current line, and autoreindent them.  It has a drawback such that before reindenting, there's a 300ms delay since `vscode-neovim.send` is executed asynchrously. I have no idea how to get rid of this.

**Comment by CAESIUS-TIM** [comment]
https://github.com/gbprod/yanky.nvim

**Comment by martin-braun** [comment]
We use vscode-neovim to profit from its built in extension ecosystem that just works. I don't really care about indent while pasting, because I just reformat things after inserting them:

```lua
    -- format buffer
    vim.keymap.set(
        'n',
        "<leader>ff",
        "<CMD>call VSCodeCall('editor.action.formatDocument')<CR>",
        { desc = "Format file" }
    )
```

I get that this is not really fulfilling when working in a codebase that you cannot format without causing massive diffs in your staged files, but this workflow generally works for me and the problem you are describing is the same deal in classic nvim, there is no auto intend while pasting.

If you want to profit from intending while inserting single lines without formatting your file, there is nobody stopping you from `o<C-v>`, instead. It won't work for multi lines, but would just bypass the problem in your provided example.

> I don't get how indent while pasting work

It does not work, because NeoVim doesn't care either, so it's good that it does not work, imho.
